crates/bevy_asset/src/query.rs

```diff
@@ -0,0 +1,165 @@
+//! Filter query to limit iteration over [`Handle`]s which content was updated recently.
+use std::marker::PhantomData;
+
+use bevy_ecs::{
+    archetype::{Archetype, ArchetypeComponentId},
+    change_detection::MAX_CHANGE_AGE,
+    component::ComponentId,
+    prelude::World,
+    query::{
+        Access, ComponentIdState, Fetch, FetchState, FilteredAccess, QueryItem, ReadFetch,
+        ReadOnlyWorldQuery, WorldQuery, WorldQueryGats,
+    },
+    storage::{Table, Tables},
+};
+use bevy_utils::HashMap;
+
+use crate::{Asset, Assets, Handle, HandleId};
+
+struct AssetChangeCheck<'w> {
+    handle_change_map: Option<&'w HashMap<HandleId, u32>>,
+    prev_system_tick: u32,
+    current_tick: u32,
+}
+impl<'w> AssetChangeCheck<'w> {
+    fn new<T: Asset>(assets: Option<&'w Assets<T>>, last_change: u32, current: u32) -> Self {
+        let handle_change_map = assets.map(|a| &a.change_ticks);
+        Self {
+            handle_change_map,
+            prev_system_tick: last_change,
+            current_tick: current,
+        }
+    }
+    fn ticks_since(&self, event_tick: u32) -> u32 {
+        self.current_tick
+            .wrapping_sub(event_tick)
+            .min(MAX_CHANGE_AGE)
+    }
+    // TODO: some sort of caching? Each check has two levels of indirection,
+    // which is not optimal.
+    fn has_changed<T: Asset>(&self, handle: &Handle<T>) -> bool {
+        let is_changed = || {
+            let map = self.handle_change_map?;
+            let handle_change_tick = *map.get(&handle.id)?;
+            // This replicates the logic used in `ComponentTicks::is_changed`
+            let ticks_since_change = self.ticks_since(handle_change_tick);
+            let ticks_since_system = self.ticks_since(self.prev_system_tick);
+            Some(ticks_since_system >= ticks_since_change)
+        };
+        is_changed().unwrap_or(true)
+    }
+}
+
+/// Query filter to get all entities which `Handle<T>` component underlying asset changed.
```

Comment:
This could use some more elaboration. In particular, I would note:

1. This is for the asset that that `Handle<T>` points to.
2. Entities without a `Handle<T>` component will never be found by this filter.
3. A small doc example :)
4. Contrast to the `Changed<Handle<T>>` query filter, which instead asks "has what the handle pointed to changed".
5. Demonstrate the `Or<(Changed<Handle<T>>, AssetChanged<T>)>` pattern, which will be commonly needed when you need to respond to asset changes.

Reply:
Yes to all this, I'll clean up the docs. In particular, I think (5) is indeed going to be very common. I'm thinking maybe a type alias could help
```rust
pub type AssetOrHandleChanged<T> = Or<(
    Changed<Handle<T>>,
    AssetChanged<T>,
)>
```
(6) is kinda misleading. I think I might just store the `change_ticks` field in its own resource, so that the incompatibility is more visible and isolated.

Reply:
I like the idea of a type alias for 5 :)